lib/puppet/functions/any_resources_of_type.rb

```diff
@@ -0,0 +1,23 @@
+Puppet::Functions.create_function(:any_resources_of_type, Puppet::Functions::InternalFunction) do 
+  dispatch :any_resources_of_type do
+    scope_param
+
+    required_param 'String', :resource_type
+    optional_param 'Hash', :parameters
+  end
+
+  def any_resources_of_type(scope, resource_type, parameters = nil)
+    scope.catalog.resources.any? do |resource|
+      # For some reason, resources types in this catalog are capitalized strings
+      # When walking the catalog on the agent, types are lowercase symbols
+      if resource.type.casecmp?(resource_type.capitalize)
```

Comment:
Figured out why this happens. When iterating over the resources on the server, we iterate over instances of `Puppet::Parser::Resource`. When doing the same thing on the agent, we get `Puppet::Type::SomeType` instances.

Aside from `resource.type` being capitalized in `Puppet::Parser::Resource` objects, parameters also differ.
`Puppet::Type` objects also have the `:name` parameter which can be accessed using `resource[:name]`, whereas `Resource` objects do not.

However, calling `Puppet::Parser::Resource#to_hash` produces a hash containing `:name` which gave me some headaches when testing this function with multiple parameters, as I was expecting `:name` to be accessible the same way as `:path`.

I wonder if there's any particular reason why we don't have a `:name` parameter on `Puppet::Resource` objects. Rediscovered this comment https://github.com/puppetlabs/puppet/pull/8174/files#r437060949 but it specifically talks about `Puppet::Type`.

Reply:
For reference, this is where `:name` gets merged in with the rest of the parameters when calling `Puppet::Parser::Resource#to_hash`:
https://github.com/puppetlabs/puppet/blob/a54094d20a52f4b0a9f2317bfb67934f9d46f8c1/lib/puppet/resource.rb#L623